### ISSUE TITLE
Redirect / rename scripts?

### ISSUE THREAD

**Issue by GavanWilhite** [vercel/analytics#102]
So, Vercel Analytics gets blocked by popular adblockers, is there a way to do something similar to Plausible and redirect / reconfigure the name of the script?

https://plausible.io/docs/proxy/guides/nextjs

**Comment by ts-web** [comment]
Yes, I set up `@vercel/analytics` in my Next.js site , but uBlock Origin blocked it, probably because of the word "insights". 

It seems like the `rewrites` feature can mask/proxy the `_vercel/insights/script` url [as described here](https://github.com/vercel/next.js/issues/22756#issuecomment-792275141), but the `Analytics` component seems to have no option to configure the URL for where it sends the analytics. Can the team add that option?

**Comment by tobiaslins** [comment]
We're automatically rewriting analytics to the 1st party intake on the same domain. After almost a year they started blocking those requests in some adblockers. We don't support custom rewrites yet. more on that topic hopefully 🔜

**Comment by TotomInc** [comment]
Sorry to resurrect this old issue but, @tobiaslins could you please explain a bit more about this?

I was wondering if it was now possible to rewrite/redirect Vercel Analytics endpoints, using Next.js redirects config.

I'm currently using PostHog as an alternative, which supports redirects with Next.js and seems to works well, but I would really prefer to stay on Vercel Analytics if you provide us the redirect alternative.

**Comment by ccbikai** [comment]
Look here, I found the perfect solution.

https://dev.to/ccbikai/jie-jue-vercel-analytics-bei-adblock-ping-bi-wen-ti-4mkn

**Comment by RusseII** [comment]
@TotomInc @ts-web @GavanWilhite 

I got this working with nextjs, but should work with any framework. 

CustomAnalytics.tsx 
```typescript
  const isDev = process.env.NEXT_PUBLIC_VERCEL_ENV === 'development'

  const beforeSend = (event: string) => {
   console.log('add custom beforeSend logic here')
    return event
  }

  const src = isDev
    ? 'https://va.vercel-scripts.com/v1/script.debug.js'
    : '/growth/script.js'

  if (typeof window !== 'undefined') {
    window.va?.('beforeSend', beforeSend)
  }

  return (
    <>
      <Script id="meow" strategy="afterInteractive">
        {`window.va = window.va || function () { (window.vaq = window.vaq || []).push(arguments); };`}
      </Script>
      <Script src={src} data-endpoint="/growth" strategy="lazyOnload" async />
    </>
  )
}
```

in next.config.js
```javascript
  async rewrites() {
    return [
      {
        source: '/growth/:match*',
        destination: '/_vercel/insights/:match*',
      },
    ]
  }
```

Enjoy! LMK if you have any problems getting it working, happy 2 help out !

**Comment by FedeMontanari** [comment]
@RusseII 

I've been trying to wrap my head around this issue, searched all over and can't seem to make it work for me.

Can you explain a bit more on how your example works?

Is the CusomAnalytics.tsx supposed to return a usable component? If so where is this component supposed to be injected?

Some analytics solutions/providers do something very similar to your proposed solution like Posthog but my knowledge is very limited to implement it in my own way.

Thanks in advance!

**Comment by tomtitherington** [comment]
I tried the approach described by both @ccbikai and @RusseII but was still blocked by the client.

Custom analytics component that I import into my root layout.
```
import Script from "next/script";

const AnalyticsEmbedder = () => {
  return (
    <>
      <Script id="ticker" strategy="afterInteractive">
        {`window.va = window.va || function () { (window.vaq = window.vaq || []).push(arguments); };`}
      </Script>
      <Script
        src="/ticker/script.js"
        data-endpoint="/ticker"
        strategy="lazyOnload"
        async
      />
    </>
  );
};

export default AnalyticsEmbedder;
```

My `next.config.mjs` file.
```
/** @type {import('next').NextConfig} */
const nextConfig = {
  trailingSlash: true,
  async rewrites() {
    return [
      {
        source: "/ticker/:path*/",
        destination: "https://mortimerhq.com/_vercel/insights/:path*/",
      },
    ];
  },
};

export default nextConfig;
```

This results in a 308 Permanent Redirect from the GET request to `https://www.mortimerhq.com/ticker/script.js`. Maybe I've misunderstood, but I thought that rewrites were not meant to expose the destination URL ... am I doing something wrong? 



For reference, I'm using `trailingSlash: true` after reading [this issue](https://github.com/vercel/next.js/discussions/36219#discussioncomment-4167863) but it makes no difference.

**Comment by nktnet1** [comment]
> Look here, I found the perfect solution.
> 
> https://dev.to/ccbikai/jie-jue-vercel-analytics-bei-adblock-ping-bi-wen-ti-4mkn

This worked for me for a **vanilla React** application, bypassing uBlock Orign.

The TLDR is:

- set a custom basePath for the `<Analytics />`
- ensure the destination URL is absolute (with protocol, domain, path) instead of relative (only path)

---

Here is the full diff:

<img src="https://github.com/user-attachments/assets/2417f936-13b1-4d6c-a1df-0a25aa922086" width="50%">


and result:

<img src="https://github.com/user-attachments/assets/fb06b7f6-7e22-4946-a849-f8f56c4cd893" width="50%">

---

<details>
<summary>Longer instructions (click to view)</summary>

My repository is:
- https://github.com/nktnet1/rt-stack

In `./apps/web/src/main.ts`, I added

```tsx
// NOTE: these imports are for a vanilla react app - import from `/next` for NextJS
import { Analytics } from '@vercel/analytics/react';
import { SpeedInsights } from '@vercel/speed-insights/react';

// ...


if (!rootElement.innerHTML) {
  const root = ReactDOM.createRoot(rootElement);
  root.render(
    <React.StrictMode>
      { /* other stuff */}


      { /* You can replace /monitor with anything else, if /monitor is also blocked */}
      <Analytics basePath="/monitor" />
      <SpeedInsights basePath="/monitor" />


      { /* other stuff */}
    </React.StrictMode>,
  );
}
```

Then, in `./apps/web/vercel.json`, I added

```jsonc
{
  "rewrites": [
    // Add this section - ensure that the destination is an absolute URL
    {
      "source": "/monitor/:match*",
      "destination": "https://rtstack.nktnet.uk/_vercel/:match*"
    },

    // This is for SPA - not important
    {
      "source": "/(.*)",
      "destination": "/index.html"
    }
  ]
}


```

</details>
